File: behavioral/null-object.go
package main

import "fmt"

type Logger interface {
	Log(msg string)
}

type FileLogger struct{}

func (FileLogger) Log(msg string) { fmt.Println("[FILE]", msg) }

type NullLogger struct{}

func (NullLogger) Log(msg string) { /* do nothing */ }

type UserService struct{ logger Logger }

func (u UserService) Register(name string) {
	fmt.Println("Registering", name)
	u.logger.Log("User " + name + " registered")
}

func main() {
	s1 := UserService{logger: FileLogger{}}
	s1.Register("Ali")

	s2 := UserService{logger: NullLogger{}}
	s2.Register("Sara")
}

File: behavioral/null-object.php
<?php
interface Logger {
    public function log(string $message): void;
}

class FileLogger implements Logger {
    public function log(string $message): void {
        echo "[FILE] $message\n";
    }
}

class NullLogger implements Logger {
    public function log(string $message): void {
        // هیچ کاری نمی‌کند
    }
}

class UserService {
    private Logger $logger;
    public function __construct(Logger $logger){
        $this->logger = $logger;
    }
    public function register(string $name){
        echo "Registering $name...\n";
        $this->logger->log("User $name registered");
    }
}

// استفاده
$service1 = new UserService(new FileLogger());
$service1->register("Ali");

$service2 = new UserService(new NullLogger());
$service2->register("Sara"); // بدون لاگ ولی بدون خطا
